### PR TITLE
Add grouped convolution support

### PR DESCRIPTION
**Pull request by tapspatel**

### CODE REVIEW

**Comment by tapspatel** [comment]
Issue: https://github.com/tenstorrent/tt-metal/issues/8340

**Comment by mywoodstock** [comment]
@tapspatel I think we should wait a couple days -- to get the new c++ convs in and then update this PR to use the new version.

**Comment by tapspatel** [comment]
pipeline: https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp [hunk 1]
```diff
@@ -881,6 +881,23 @@ void TensorModule(py::module& m_tensor) {
...
+        Converts convolution weights to grouped layout with padded zeros
+        Returns a new tensor with the converted layout.
+
+        +----------+----------------------+-----------+-------------+----------+
+        | Argument | Description          | Data type | Valid range | Required |
+        +==========+======================+===========+=============+==========+
+        | a        | Input tensor         | Tensor    |             | Yes      |
+        +----------+----------------------+-----------+-------------+----------+
+    )doc");
+
```
Comment (TT-BrianLiu) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp:900]: I think we should eventually move these bindings to a lower level module that's not just `ttl.tensor.*`.

**Review by TT-BrianLiu — APPROVED** [review]

**Comment by mywoodstock** [comment]
@tapspatel could you please use the new `test_new_conv2d.py` for the latest version?

**Review by cfjchu — APPROVED** [review]

File: tt_eager/tensor/tensor_utils.cpp [hunk 2]
```diff
@@ -402,68 +534,81 @@ Tensor get_shard_for_device(const Tensor& tensor, Device* target_device, std::op
...
     // They need to be copied to owned storage before being passed to the worker.
     ZoneScopedN("ConvertBorrowedToOwned");
     // Tensor has workers (on device) or runtime mode is synchronous or tensor has multiple buffers.
     // No need to check for borrowed storage.
-    if (worker->get_worker_mode() == WorkExecutorMode::SYNCHRONOUS or
-        tensor.get_workers().size() or
-        tensor.tensor_attributes->tensor_populated.size() > 1) return tensor;
+    if (worker->get_worker_mode() == WorkExecutorMode::SYNCHRONOUS or tensor.get_workers().size() or
+        tensor.tensor_attributes->tensor_populated.size() > 1)
+        return tensor;
```
Comment (xanderchin) [inline tt_eager/tensor/tensor_utils.cpp:593]: Can we always brace bracket even single line ifs?

**Review by xanderchin — APPROVED** [review]

File: tt_eager/tensor/tensor_utils.cpp (hunk 2 above)
Comment (tapspatel) [reply tt_eager/tensor/tensor_utils.cpp:593]: fixed

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp [hunk 3]
```diff
@@ -881,6 +881,23 @@ void TensorModule(py::module& m_tensor) {
...
+        py::arg("num_groups"),
+        py::arg("output_dtype").noconvert() = std::nullopt,
+        R"doc(
+        Converts convolution weights to grouped layout with padded zeros
+        Returns a new tensor with the converted layout.
+
+        +----------+----------------------+-----------+-------------+----------+
+        | Argument | Description          | Data type | Valid range | Required |
+        +==========+======================+===========+=============+==========+
+        | a        | Input tensor         | Tensor    |             | Yes      |
```
Comment (tt-aho) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp:897]: This and some of the other bound conv functions needs a proper arg breakdown/description.

**Review by tt-nshanker — APPROVED** [review]

**Comment by tapspatel** [comment]
pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/9356946489

**Review by mywoodstock — APPROVED** [review]

**Pull request merged**